tests/test_sdeint.py

```diff
@@ -35,152 +36,149 @@
+@pytest.mark.parametrize('adaptive', [False, True])
+@pytest.mark.parametrize('logqp', [False, True])
+@pytest.mark.parametrize('device', devices)
+def test_sdeint_run_shape_method(sde_cls, use_bm, levy_area_approximation, sde_type, method, adaptive, logqp, device):
+    # Tests that sdeint:
+    # (a) runs/raises an error as appropriate
+    # (b) produces tensors of the right shape
+    # (c) accepts every method
+
+    if logqp:
+        pytest.skip("Temporarily deprecating logqp.")
+
+    should_fail = False
+    if sde_type == 'ito':
+        if method in ('euler_heun', 'heun', 'midpoint', 'log_ode'):
+            should_fail = True
+    else:
+        if method in ('euler', 'srk'):
+            should_fail = True
+    if method in ('milstein', 'srk') and sde_cls.noise_type == 'general':
+        should_fail = True
+    if method == 'srk' and levy_area_approximation == 'none':
+        should_fail = True
+    if method == 'log_ode' and levy_area_approximation in ('none', 'space-time'):
+        should_fail = True
+    if method == 'blah':
+        should_fail = True
+
+    if sde_cls is basic_sde.ScalarSDE:
+        kwargs = {'d': d}
+    elif sde_cls is basic_sde.DiagonalSDE:
+        kwargs = {}
+    else:
+        kwargs = {'d': d, 'm': m}
+    sde = sde_cls(sde_type=sde_type, **kwargs).to(device)
+
+
+    if use_bm:
+        if sde_cls.noise_type == 'scalar':
+            size = (batch_size, 1)
+        elif sde_cls.noise_type == 'diagonal':
+            size = (batch_size, d)
+        else:
+            assert sde_cls.noise_type in ('additive', 'general')
+            size = (batch_size, m)
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, size=size, dtype=dtype, device=device,
+                                       levy_area_approximation=levy_area_approximation)
+    else:
+        bm = None
+
+    _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail)
+
+
+@pytest.mark.parametrize("sde_cls", [basic_sde.BasicSDE1, basic_sde.BasicSDE2, basic_sde.BasicSDE3,
+                                     basic_sde.BasicSDE4])
+@pytest.mark.parametrize('method', ['euler', 'milstein', 'srk'])
+@pytest.mark.parametrize('adaptive', [False, True])
+@pytest.mark.parametrize('device', devices)
+def test_sdeint_dependencies(sde_cls, method, adaptive, device):
+    # This test uses diagonal noise. This checks if the solvers still work when some of the functions don't depend on
+    # the states/params and when some states/params don't require gradients.
+    sde = sde_cls(d=d).to(device)
+    bm = None
+    logqp = False
+    should_fail = False
+    _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail)
+
+
+def _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail):
+    y0 = torch.ones(batch_size, d, device=device)
+    ts = torch.linspace(t0, t1, steps=T, device=device)
+    if adaptive and method == 'euler' and sde.noise_type != 'additive':
```

Comment:
Doing this userwarning check here is probably going a bit far. This if-check is basically the exact same code as in `sdeint.py`, so I'd think it's unlikely something will go wrong. I wouldn't recommend including it just to keep things simple and easily maintainable. Not a super strong feeling though.

Reply:
This is mostly just so that the warning doesn't bubble up to get reported at the pytest level.
(I originally had a `warnings.simplefilter` on this, but figured if I was going to include a context like that then I might as well do it properly.)